src/sys/windows/named_pipe.rs

```diff
@@ -83,21 +64,43 @@ pub struct NamedPipe {
     inner: Arc<Inner>,
 }
 
+/// # Notes
+///
+/// The memory layout of this structure must be fixed as the `*_OFFSET`
+/// constants depend on it, see the `offset_constants` test.
 #[repr(C)]
 struct Inner {
-    handle: pipe::NamedPipe,
-
     connect: Overlapped,
-    connecting: AtomicBool,
-
     read: Overlapped,
     write: Overlapped,
-
+    handle: pipe::NamedPipe,
+    connecting: AtomicBool,
     io: Mutex<Io>,
-
     pool: Mutex<BufferPool>,
 }
 
+/// Offsets from a pointer to `Inner` to the `connect`, `read` and `write`
+/// fields.
+const CONNECT_OFFSET: usize = 0;
+const READ_OFFSET: usize = CONNECT_OFFSET + size_of::<Overlapped>(); // `connect` fields.
+const WRITE_OFFSET: usize = READ_OFFSET + size_of::<Overlapped>(); // `read` fields.
```

Comment:
This seems rather fragile...

If I understand correctly, we actually do have an instance of the relevant type available at the place where the offset needs to be computed. So why can't the code simply compute the offsets based on that instance? `offset_of!` is only hard when you don't have a value of the type you want to compute offsets in.

Reply:
I agree it's fragile.

But we don't have an instance available in [`read_done`](https://github.com/tokio-rs/mio/pull/1485/files#diff-8648c6b3c0e6fc7faf5a29bd380248e904852cb3bfaf6e0e2638af9d281b4960R614) when we need it. We have a pointer to one of the fields of the instance. How do you suggest we get the base pointer (to `Inner`) then?

Reply:
Ah, I thought you'd have an instance anywhere, sorry... I don't really understand the code.^^

Btw, this looks like `container_of!`-style pointer arithmetic, which does not work well with the Rust aliasing model... https://github.com/rust-lang/unsafe-code-guidelines/issues/243.

But ignoring aliasing issues, if you do not want to depend on `addr_of!`, you could either decide to not care about code being technically UB (you could still avoid the NULL ptr by using `NonNull::dangling` as the base) -- or I guess hard-code the offsets as you do...

Reply:
I don't really know of a way to solve this. I'm following https://github.com/rust-lang/unsafe-code-guidelines/issues/243 now so if someone suggests a solution that is not UB we can always update this code.